chore: add hardware wallet events

gantunesr opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/37717?quickstart=1)

This PR adds a set of hardware wallet events necessary to build a connection funnel dashboard and find drops to improve the UX.

Events added:

- `Add Hardware Wallet Clicked`
- `Connect Hardware Wallet Clicked`
- ~`Connect Hardware Wallet Device Found`~
- ~`Connect Hardware Wallet Continue Button Clicked`~
- `Connect Hardware Wallet Account Selector Viewed`
- `Hardware Wallet Account Connected`
- `Hardware Wallet Forgotten`
- `Hardware Wallet Marketing Button Clicked`
- `Hardware Wallet Connection Failed`

## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and accurately)
-->

CHANGELOG entry: null

## **Related issues**

Fixes: https://consensyssoftware.atlassian.net/browse/MUL-1240

## **Manual testing steps**

1. Setup a Segment account as indicated in the [README](https://github.com/MetaMask/metamask-extension/blob/4ce369dea01a3ec657b7c78950f282bb33edda26/development/README.md#L7)
2. Connect a hardware wallet
3. Check Segment dashboards for the hardware wallet events

## **Screenshots/Recordings**

<img width="1695" height="834" alt="Screenshot 2025-11-12 at 6 23 10 PM" src="https://github.com/user-attachments/assets/5b4fcfe2-1b61-4fd8-b514-f1341ecd747f" />

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.



<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Adds new MetaMetrics events for the hardware wallet connection funnel and wires tracking across the add wallet modal and connect hardware flow, plus constants/util updates.
> 
> - **MetaMetrics (events/constants)**
>   - Add hardware wallet events in `MetaMetricsEventName` (e.g., `AddHardwareWalletClicked`, `ConnectHardwareWalletClicked`, `HardwareWalletAccountConnected`, `HardwareWalletForgotten`, `HardwareWalletMarketingButtonClicked`, `HardwareWalletConnectionFailed`, etc.).
>   - Export `MetaMetricsEventPayload` usage in UI.
> - **Shared constants (`shared/constants/hardware-wallets.ts`)**
>   - Add `QrHardwareDeviceNames`, `MarketingActionNames`, `DEVICE_KEYRING_MAP`, `LEDGER_ERRORS_CODES`, and standard HD path constants (`LEDGER_LIVE_PATH`, `MEW_PATH`, `BIP44_PATH`, Lattice/Trezor variants).
>   - Normalize affiliate links/tutorial links to PascalCase keys in `HardwareAffiliateLinks` and `HardwareAffiliateTutorialLinks`.
> - **UI: Add Wallet Modal**
>   - Track `AddHardwareWalletClicked` when selecting hardware wallet in `add-wallet-modal.tsx`.
> - **UI: Connect Hardware flow**
>   - `select-hardware.js`: track marketing button clicks with `HardwareWalletMarketingButtonClicked`; use new affiliate link enums; add USB/Trezor pre-request; capitalize device names via `capitalizeStr`.
>   - `index.js`: track funnel steps (`ConnectHardwareWalletClicked`, `HardwareWalletAccountConnected`, `HardwareWalletConnectionFailed`, `HardwareWalletForgotten`); include properties like `device_type`, `connected_device_count`, `hd_path`; refactor error mapping to `LEDGER_ERRORS_CODES`; add keyring introspection for connected count; pass `keyrings` from state.
>   - `account-list.js`: track `ConnectHardwareWalletAccountSelectorViewed`; pass `hdPath` to forget; use `capitalizeStr`.
>   - Add `utils.ts` with `capitalizeStr` helper.
> - **Storybook**
>   - Update stories to mock `trackEvent` context for `AccountList` and `SelectHardware`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 197b1a3df981ed5538abe665876e406e98e9b652. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

gantunesr committed:
chore: add event names

gantunesr committed:
chore: add constant values

gantunesr committed:
chore: add events

gantunesr committed:
Merge branch 'main' of https://github.com/MetaMask/metamask-extension into gar/chore/hw-events

gantunesr committed:
add event

gantunesr committed:
add missing line

gantunesr committed:
chore: add category

gantunesr committed:
chore: add device count

gantunesr committed:
chore: delete old events

gantunesr committed:
Merge branch 'main' of https://github.com/MetaMask/metamask-extension into gar/chore/hw-events

mathieuartu commented on ui/pages/create-account/connect-hardware/select-hardware.js:
```diff
@@ -657,10 +684,10 @@ export default class SelectHardware extends Component {
-                });
+                this.trackMarketingEvent(
+                  MarketingActionNames.Tutorial,
+                  QrHardwareDeviceNames.keystone,
```
@montelaidev seems like this comment is correct

mathieuartu reviewed (commented):
LGTM, a couple of casing issues

gantunesr committed:
fix: enum names

gantunesr committed:
fix: enum names

gantunesr commented on ui/pages/create-account/connect-hardware/select-hardware.js:
```diff
@@ -692,10 +719,10 @@ export default class SelectHardware extends Component {
-                });
+                this.trackMarketingEvent(
+                  MarketingActionNames.Tutorial,
+                  QrHardwareDeviceNames.airgap,
```
solved at [f98da20](https://github.com/MetaMask/metamask-extension/pull/37717/commits/f98da209e63419807283a98179d2971c9802c635)

gantunesr committed:
Merge branch 'main' into gar/chore/hw-events

gantunesr committed:
fix: change method to async

gantunesr committed:
Merge branch 'gar/chore/hw-events' of https://github.com/MetaMask/metamask-extension into gar/chore/hw-events

gantunesr committed:
Merge branch 'main' into gar/chore/hw-events

owencraston commented:
# Trezor Testing Results

## Overall
- The `device_type` is set to `trezor` when it should be `Trezor` to match mobile and the definitions we defined in the segment schema
- No Connect Hardware Wallet Device Found (same comment/question as Ledger): This event is not being fired when the device becomes available to MetaMask (when it appears in the chrome popup). Based on previous discussions I know we had issues with this step since we do not own the browser portion. Let me know if this is/is not possible.
-connected_device_count:
    - This property is missing from the Connect Hardware Wallet Clicked event and should default to 0 if no hardware wallets are connected
    - The value of this property seems to be set correctly with Trezor but not with Ledger. Not exactly sure what the difference in logic is here. 
- missing device_model: We discussed that this would come in a follow up PR so not an issue. I just wanted to call it out for the sake of clarity.

Below is a summary of the events that fired during my testing:


### Connection flow

```
[mock-segment]: Track event received: Add Hardware Wallet Clicked
{
  "category": "Navigation",
  "locale": "en",
  "chain_id": "0xaa36a7",
  "environment_type": "popup"
}
[mock-segment]: Page event received: Connect Hardware Wallet Page
{
  "params": {},
  "locale": "en",
  "chain_id": "0xaa36a7",
  "environment_type": "fullscreen"
}

[mock-segment]: Track event received: Connect Hardware Wallet Clicked
{
  "device_type": "trezor",
  "locale": "en",
  "chain_id": "0xaa36a7",
  "environment_type": "fullscreen"
}
[mock-segment]: Track event received: Connect Hardware Wallet Continue Button Clicked
{
  "device_type": "trezor",
  "locale": "en",
  "chain_id": "0xaa36a7",
  "environment_type": "fullscreen"
}
[mock-segment]: Track event received: Connect Hardware Wallet Account Selector Viewed
{
  "device_type": "trezor",
  "locale": "en",
  "chain_id": "0xaa36a7",
  "environment_type": "fullscreen"
}

[mock-segment]: Track event received: Hardware Wallet Account Connected
{
  "device_type": "trezor",
  "hd_path": "m/44'/60'/0'/0",
  "connected_device_count": 1,
  "locale": "en",
  "chain_id": "0xaa36a7",
  "environment_type": "fullscreen"
}
```

### Marketing buttons ✅ 

```
[mock-segment]: Track event received: Hardware Wallet Marketing Button Clicked
{
  "button_type": "Tutorial",
  "device_type": "trezor",
  "locale": "en",
  "chain_id": "0xaa36a7",
  "environment_type": "fullscreen"
}
[mock-segment]: Track event received: Hardware Wallet Marketing Button Clicked
{
  "button_type": "Buy Now",
  "device_type": "trezor",
  "locale": "en",
  "chain_id": "0xaa36a7",
  "environment_type": "fullscreen"
}
```

### Forgetting a device

```
[mock-segment]: Track event received: Hardware Wallet Forgotten
{
  "device_type": "trezor",
  "locale": "en",
  "chain_id": "0xaa36a7",
  "environment_type": "fullscreen"
}
```

gantunesr committed:
chore: solve review comments

gantunesr committed:
Merge branch 'main' into gar/chore/hw-events

gantunesr committed:
refactor: update affiliate link support

gantunesr committed:
Merge branch 'gar/chore/hw-events' of https://github.com/MetaMask/metamask-extension into gar/chore/hw-events

gantunesr committed:
chore: remove dupe event

gantunesr committed:
chore: capitalize device name

gantunesr committed:
fix: count

gantunesr committed:
fix: event trigger

gantunesr committed:
fix: solve extra count

gantunesr committed:
fix: solve extra count

gantunesr committed:
fix: lint

gantunesr committed:
chore: remove try-catch

gantunesr committed:
chore: improve comment

gantunesr committed:
chore: add failed event

gantunesr committed:
chore: update failed event

gantunesr committed:
fix: update events

gantunesr committed:
fix: remove event

gantunesr commented on shared/constants/hardware-wallets.ts:
```diff
@@ -160,4 +131,37 @@ export enum WebHIDConnectedStatuses {
+  [HardwareDeviceNames.trezor]: KeyringTypes.trezor,
+  [HardwareDeviceNames.oneKey]: KeyringTypes.oneKey,
+  [HardwareDeviceNames.lattice]: KeyringTypes.lattice,
+  [HardwareDeviceNames.qr]: KeyringTypes.qr,
```
All defined values in the map should exist in the `KeyringTypes` from the `@metamask/keyring-controller`

gantunesr committed:
chore: remove event

gantunesr committed:
chore: update storybook

gantunesr committed:
solve comments

gantunesr committed:
Merge branch 'main' into gar/chore/hw-events

gantunesr committed:
fix: lint

gantunesr committed:
Merge branch 'gar/chore/hw-events' of https://github.com/MetaMask/metamask-extension into gar/chore/hw-events

gantunesr committed:
fix: update storybook

gantunesr committed:
chore: update enum

gantunesr committed:
update event props

montelaidev committed:
fix: storybook (#38049)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

Fix storybook

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/38049?quickstart=1)

## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the
CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing
description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and
accurately)
-->

CHANGELOG entry:

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask
Extension Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> <sup>[Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) is
generating a summary for commit
11abd1f4c5addc3ea678ca1b24dce713528b837d. Configure
[here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

mathieuartu reviewed (approved):
Latest changes LGTM

gantunesr committed:
Merge branch 'main' into gar/chore/hw-events

gantunesr commented on ui/pages/create-account/connect-hardware/index.js:
```diff
@@ -81,18 +78,11 @@ const HD_PATHS = {
-    default:
-      return errorCode;
+  if (Object.keys(LEDGER_ERRORS_CODES).includes(errorCode)) {
+    return t(LEDGER_ERRORS_CODES[errorCode]);
```
https://github.com/MetaMask/metamask-extension/pull/38079
